### PR TITLE
Allow to customize the fetchPolicy with interceptors

### PR DESCRIPTION
**Pull request by martinbonnin**
Splits `ApolloCacheInterceptor` into

* `WatcherInterceptor`
* `FetchPolicyRouterInterceptor`
* `ApolloCacheInterceptor`

This allows to set interceptors for a specific FetchPolicy. See https://github.com/apollographql/apollo-kotlin/issues/3689 for more details

### CODE REVIEW

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt [hunk 1]
```diff
@@ -232,36 +260,142 @@ internal val <D : Operation.Data> ApolloRequest<D>.cacheHeaders
...
+      networkEndMillis = 0,
+      cacheHit = hit,
+      cacheMissException = missedKey?.let { CacheMissException(it, missedField) },
+      networkException = null
+  )
+
   override val key: ExecutionContext.Key<*>
     get() = Key
 
+  @Deprecated("Use cacheStartMillis instead", ReplaceWith("cacheStartMillis"))
```
Comment (BoD) [inline apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt:298]: A completely adjacent thought: could it be a good idea to keep a log of when (in which version) APIs were deprecated, if we need to reason about when they can be deleted. Kotlin has a `DeprecatedSinceKotlin` annotation. Maybe I'm overthinking and it's unnecessary though :)

Comment (martinbonnin) [reply apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt:298]: I'm always for more strict API versionning 👍

Comment (martinbonnin) [reply apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt:298]: Follow up issue: https://github.com/apollographql/apollo-kotlin/issues/3745

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/internal/WatcherInterceptor.kt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    check(request.operation is Query) {
+      "It's impossible to watch a mutation or subscription"
+    }
+
+    val customScalarAdapters = request.executionContext[CustomScalarAdapters]!!
+    var watchedKeys: Set<String>? = null
+    var isRefetching = false
+
+    return flowOf(flowOf(emptySet()), store.changedKeys)
+        .flattenConcatPolyfill()
```
Comment (BoD) [inline apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/internal/WatcherInterceptor.kt:41]: Is this equivalent (and clearer?)?

```suggestion
    return store.changedKeys.onStart { emit(emptySet()) }
```

Comment (martinbonnin) [reply apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/internal/WatcherInterceptor.kt]: 👍  much much better

**Comment by BoD** [comment]
Initial review of the draft: the code is more modular/decoupled which is very nice + makes the code easier to grasp!

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/FetchPolicyInterceptors.kt [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+import com.apollographql.apollo3.interceptor.ApolloInterceptor
+import com.apollographql.apollo3.interceptor.ApolloInterceptorChain
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.singleOrNull
+import kotlin.jvm.JvmName
+
+/**
+ *
```
Comment (BoD) [inline apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/FetchPolicyInterceptors.kt:21]:
```suggestion
 * An interceptor that goes to the cache only
```

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt [hunk 4]
```diff
@@ -232,36 +260,142 @@ internal val <D : Operation.Data> ApolloRequest<D>.cacheHeaders
...
+ * @param cacheHit true if this was a cache hit
+ * @param cacheException the exception while reading the cache. Note that it's possible to have [cacheHit] == false && [cacheException] == null
+ * if no cache read was attempted
+ */
+class CacheInfo private constructor(
+    val cacheStartMillis: Long,
+    val cacheEndMillis: Long,
+    val networkStartMillis: Long,
+    val networkEndMillis: Long,
+    val cacheHit: Boolean,
```
Comment (BoD) [inline apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt:273]: Nitpick: `isCacheHit`?

**Review by BoD — APPROVED** [review]
👍

File: apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt (hunk 4 above)
Comment (martinbonnin) [reply apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt:273]: 👍

**Pull request merged**